### ISSUE TITLE
Stack-overflow occurring on 1.0.x (Works in 0.10.x)

### ISSUE THREAD

**Issue by markvrensburg** [functional-streams-for-scala/fs2#1335]
Here is a minimal rose-tree implementation using fs2:
```
import fs2.{Pure, Stream}

case class Tree[A](label: A, subForest: Stream[Pure,Tree[A]]) {
  /** Depth-first pre-order traversal */
  def flatten: Stream[Pure, A] =
    Stream(this.label) ++ this.subForest.flatMap(_.flatten)
}

object Tree {
  def unfoldForest[A, B](s: Stream[Pure,A])(f: A => (B, Stream[Pure,A])): Stream[Pure,Tree[B]] =
    s.map(unfoldTree(_)(f))

  def unfoldTree[A, B](seed: A)(f: A => (B, Stream[Pure, A])): Tree[B] =
    f(seed) match {
      case (a, bs) => Tree(a, unfoldForest(bs)(f))
    }
}
```

This behaves as expected in `1.0.x`:
```
//         1
//       /   \
//      2     2
//     /|\   /|\
//    3 3 3 3 3 3
//    : : : : : :

val tree1 = Tree.unfoldTree(1)(x => (x, Stream(x+1).repeat.take(x+1)))
```

But this stack-overflows on `1.0.x` but works in `0.10.x`:

```
//    1
//    |
//    2
//    |
//    3
//    :

val tree2 = Tree.unfoldTree(1)(x => (x, Stream(x+1)))
```

A work around in `1.0.x` is something like:
`val tree2 = Tree.unfoldTree(1)(x => (x, Stream(x+1) ++ Stream.empty))`

Adding `++ Stream.empty` seems a bit weird?

https://scalafiddle.io/sf/oi8aWb1/0

**Comment by mpilquist** [comment]
Here's a representative part of the stack trace:

```
  fs2.Chunk$$anon$5.map(Chunk.scala:481)
  fs2.Pull$$anon$3.apply(Pull.scala:187)
  fs2.Pull$$anon$3.apply(Pull.scala:185)
  fs2.internal.FreeC$Eval.translate(FreeC.scala:197)
  fs2.Pull$.mapOutput(Pull.scala:185)
  fs2.Pull$.mapOutput$extension(Pull.scala:82)
  fs2.Stream$.map$extension(Stream.scala:1348)
  ammonite.$sess.cmd3$Tree$.unfoldForest(cmd3.sc:9)
  ammonite.$sess.cmd3$Tree$.unfoldTree(cmd3.sc:13)
  ammonite.$sess.cmd3$Tree$.$anonfun$unfoldForest$1(cmd3.sc:9)
  fs2.Chunk$$anon$5.map(Chunk.scala:481)
  fs2.Pull$$anon$3.apply(Pull.scala:187)
  fs2.Pull$$anon$3.apply(Pull.scala:185)
  fs2.internal.FreeC$Eval.translate(FreeC.scala:197)
  fs2.Pull$.mapOutput(Pull.scala:185)
  fs2.Pull$.mapOutput$extension(Pull.scala:82)
  fs2.Stream$.map$extension(Stream.scala:1348)
  ammonite.$sess.cmd3$Tree$.unfoldForest(cmd3.sc:9)
  ammonite.$sess.cmd3$Tree$.unfoldTree(cmd3.sc:13)
  ammonite.$sess.cmd3$Tree$.$anonfun$unfoldForest$1(cmd3.sc:9)
```

`unfoldForest` calls `Stream#map` which eagerly calls `translate` on the underlying `FreeC` which eagerly calls the supplied `FunctionK` which eagerly calls `o.values.map(f)`.

We could fix by making `FreeC#translate` suspend.

**Comment by markvrensburg** [comment]
Thanks @mpilquist for taking a look at this. I wish I knew a bit more about the fs2 internals to have helped you out.

**Comment by mpilquist** [comment]
Here's a smaller reproduction - just inlined a bunch of stuff to make the problematic structure easier to spot:

```scala
case class Tree[A](label: A, subForest: Stream[Pure, Tree[A]]) {
  def flatten: Stream[Pure, A] =
    Stream(this.label) ++ this.subForest.flatMap(_.flatten)
 }

def unfoldTree(seed: Int): Tree[Int] =
  Tree(seed, Stream(seed + 1).map(unfoldTree))

unfoldTree(1).flatten.take(10).toList shouldBe List.tabulate(10)(_ + 1)
```

I'm not *exactly* sure what the issue is here but it's related to the fact that `map` is now implemented with `translate` and `translate` is eager when the "root" of the stream is an `Output` node. Hence, one way to fix/workaround this issue is suspending the stream that gets mapped:

```scala
case class Tree[A](label: A, subForest: Stream[Pure, Tree[A]]) {
  def flatten: Stream[Pure, A] =
    Stream(this.label) ++ this.subForest.flatMap(_.flatten)
 }

def unfoldTree(seed: Int): Tree[Int] =
  Tree(seed, Stream.suspend(Stream(seed + 1)).map(unfoldTree))

unfoldTree(1).flatten.take(10).toList shouldBe List.tabulate(10)(_ + 1)
```

**Comment by mpilquist** [comment]
Here's a clearer example that shows the main problem here is unguarded recursion of `unfoldTree` -- here I removed the call to `map` completely.

```scala
case class Tree[A](label: A, subForest: Stream[Pure, Tree[A]]) {
  def flatten: Stream[Pure, A] =
    Stream(this.label) ++ this.subForest.flatMap(_.flatten)
}

def unfoldTree(seed: Int): Tree[Int] =
  Tree(seed, Stream(unfoldTree(seed + 1)))

unfoldTree(1).flatten.take(10).toList shouldBe List.tabulate(10)(_ + 1)
```

**Comment by mpilquist** [comment]
One more example to show that `map` is strict if the root / prefix of the stream is an `Output` node (e.g. from `emit`, `emits`, `apply`):

```scala
@ Stream(1, 2, 3).map { i => println("Mapping " + i); i }
Mapping 1
Mapping 2
Mapping 3
res1: Stream[Nothing, Int] = Stream(..)
```

The question here really is whether the performance gains are worth the confusion that might be caused as a result of a strict operation. I tend to think the current implementation is correct and this use case needs a `Stream.suspend` call to guard the recursion.

Interested in opinions. @pchlupacek @SystemFw @djspiewak

**Comment by markvrensburg** [comment]
@mpilquist I'm happy with the workarounds you suggested; any help with improving the implementation is greatly appreciated. Most of my implementation is taken from Scalaz and Haskell. The more challenging issues I had was implementing the breadth-first traversal and zippers with fs2, but it has been a good learning experience.

**Comment by markvrensburg** [comment]
I'm going for this solution for this particular use case:

```
def unfold[A, B](seed: A)(f: A => (B, Stream[Pure, A])): Tree[B] =
    f(seed) match {
      case (b, as) => Tree(b, Stream.suspend(as.map(unfold(_)(f))))
    }
```

It feels the cleanest to me. Thanks very much for the insights.

**Comment by SystemFw** [comment]
> The question here really is whether the performance gains are worth the confusion that might be caused as a result of a strict operation

I think this could benefit from some numbers perhaps? At the moment I tend to agree with you that the current implementation is the right tradeoff, however I'm afraid of discoverability, how will people figure out they need to add a `suspend` when there is unguarded recursion on an Output node?

**Comment by mpilquist** [comment]
@markvrensburg We merged a better fix for this in #1336 after all -- will be in the next release, which is hopefully this week.

**Comment by markvrensburg** [comment]
Wow thanks @mpilquist, it is really so awesome to witness the fs2 contributors process of resolving these types of issues so quickly. So much to learn in open source software dev in this.

When looking back at how I presented the original problem, is there anything you could suggest that I could have done to make it it easier for you guys to investigate. ie. supplying stack traces, minimising the reproduction a little bit more, etc.
